### PR TITLE
Sandbox IDE MVP

### PR DESCRIPTION
**Pull request by calebjacob**
Closes: https://github.com/near/bos-web-engine/issues/86

- Created a very thin Next JS wrapper in `apps/sandbox` which renders our new `<Sandbox />` component provided by the new `@bos-web-engine/sandbox` package. The thought is that this separate package will allow other developers to easily include the BWE sandbox as well.
- When running `pnpm dev`, the sandbox app will start up on port `3005` - leaving `3000` to render `apps/web`.
- Updated `useWebEngine()` hook to allow passing in an optional `localComponents` array. Whenever this array or `rootComponentPath` changes, the compiler will reset the cache and re-transpile all components.
- Whenever the passed in values for `localComponents` or `rootComponentPath` change, a new `nonce` value will be returned from the hook. This is useful to attach as a `key` value to your `<ComponentTree key={nonce} />` if you plan on these values dynamically changing (like in an IDE environment). This completely resets the DOM for the `ComponentTree` and allows for a fresh slate before recompiling.
- Updated `package.json` and added `.nvmrc` to better enforce correct versions of `Node` and `pnpm`. The newer version of `pnpm` caused a full reformatting of our lock file.

Remaining Items:

- [x] Loading state/animation (while loading external libraries)
- [x] Add file
- [x] Rename file
- [x] Delete file
- [x] Save files to local storage on change
- [x] Collapse/expand editor toggle (and keyboard shortcut) to make preview take up full screen or editor take up full screen

### CODE REVIEW

File: packages/compiler/src/types.ts [hunk 1]
```diff
@@ -13,6 +14,13 @@ export interface CompilerInitAction {
   preactVersion: string;
 }
 
+export interface CompilerSetLocalComponentAction {
+  action: 'set-local-components';
+  components: { componentPath: string; componentSource: string }[];
+  resetCache?: boolean;
```
Comment (andy-haynes) [inline packages/compiler/src/types.ts:20]: 🔪

File: packages/compiler/src/types.ts [hunk 2]
```diff
@@ -13,6 +14,13 @@ export interface CompilerInitAction {
   preactVersion: string;
 }
 
+export interface CompilerSetLocalComponentAction {
+  action: 'set-local-components';
+  components: { componentPath: string; componentSource: string }[];
```
Comment (andy-haynes) [inline packages/compiler/src/types.ts:19]: Can we make this a mapping of path -> source to make its purpose more clear? As an array there's nothing preventing duplicates and setting values is more involved than upserting an entry.

**Review by andy-haynes — DISMISSED** [review]
Looking great! Just a couple comments on the new compiler interface, in particular I'd prefer we change the type for tracking local Components.

File: packages/compiler/src/types.ts (hunk 1 above)
Comment (calebjacob) [reply packages/compiler/src/types.ts:20]: Good catch! `resetCache` was leftover from my first draft and it slipped through the cracks.

File: packages/compiler/src/types.ts (hunk 2 above)
Comment (calebjacob) [reply packages/compiler/src/types.ts:19]: For sure! That's a great point. Will refactor that now.

File: apps/sandbox/README.md
Comment (mpeterdev) [inline apps/sandbox/README.md:1]: please swap out the boilerplate README with something appropriate, can even just be a single line description of the app

File: apps/sandbox/public/next.svg
Comment (mpeterdev) [inline apps/sandbox/public/next.svg:1]: kill

File: apps/sandbox/public/vercel.svg
Comment (mpeterdev) [inline apps/sandbox/public/vercel.svg:1]: kill

**Review by mpeterdev — CHANGES_REQUESTED** [review]
add a short README to `packages/sandbox` as well 🙏

File: apps/sandbox/README.md
Comment (calebjacob) [reply apps/sandbox/README.md:1]: @mpeterdev Good point! Forgot to do that. I updated both `README`'s with some basic content for now.

**Review by mpeterdev — APPROVED** [review]

**Review by mpeterdev — APPROVED** [review]

**Comment by charleslavon** [comment]
🔥

**Pull request merged**